homeassistant/components/geniushub/__init__.py

```diff
@@ -0,0 +1,38 @@
+"""This module connects to the Genius hub and shares the data."""
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_HOST, CONF_PASSWORD, CONF_USERNAME, CONF_SCAN_INTERVAL)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.discovery import async_load_platform
+
+REQUIREMENTS = ['geniushub==0.1']
+GENIUS_HUB = 'genius_hub'
+DOMAIN = 'geniushub'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=6): cv.time_period,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Try to start embedded Genius Hub broker."""
+    from geniushub.geniushub import GeniusHub
+
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    host = config.get(CONF_HOST)
+    scan_interval = config.get(CONF_SCAN_INTERVAL)
+
+    hass.data[GENIUS_HUB] = GeniusHub(
```

Comment:
Please have the client accept an (optional) aiohttp session. We want all integrations that use aiohttp in home assistant to use the aiohttp session of home assistant.

Use our aiohttp helper to get the session:
https://github.com/home-assistant/home-assistant/blob/59476ab475122bb02ddbdd2abcd4930cdfece849/homeassistant/helpers/aiohttp_client.py#L30

Reply:
Also, please have the client (optionally) accept an asyncio event loop to use. We want all asyncio integrations to use the home assistant event loop.

Reply:
@MartinHjelmare can you give an example of a component that does this?  I had a look at hue, but it uses config flow...

Reply:
Actually, I found: 
```python
async def get_bridge(hass, host, username=None):
    """Create a bridge object and verify authentication."""
    import aiohue

    bridge = aiohue.Bridge(
        host, username=username,
        websession=aiohttp_client.async_get_clientsession(hass)
    )
```

Reply:
Tibber uses our session:
https://github.com/home-assistant/home-assistant/blob/59476ab475122bb02ddbdd2abcd4930cdfece849/homeassistant/components/tibber/__init__.py#L32-L33

Reply:
I have addressed this:
https://github.com/home-assistant/home-assistant/blob/17cb347a844c6eadbc029307b2b0afe15be54979/homeassistant/components/geniushub/__init__.py#L36-L39

But, I am getting an occasional  `ServerDisconnectedError ` error in the client library.  The API is undocumented, so I have done this:
```python
    async def _request(self, method, url, data=None):
    ...    ...
        try:
            _LOGGER.warning("_request(): 1st try: %s %s %s", method, url, data)
            async with http_method(
    ...    ...
            ) as resp:
                assert resp.status == HTTP_OK
                response = await resp.json()
            return response


        except aiohttp.client_exceptions.ServerDisconnectedError as err:
            _LOGGER.warning(
                "_request(): 2nd try: %s %s %s - as ServerDisconnected. "
                "Message was: %s", method, url, data, err)
            _session = aiohttp.ClientSession()
            async with http_method(
    ...     ...
            ) as resp:
                assert resp.status == HTTP_OK
                response = await resp.json()
            await _session.close()
            return response
```

The Exception only occurs with something other than a `GET`.

Reply:
see: https://github.com/zxdavb/geniushub-client/blob/6e7b038e33de7b4a84b74b4a243c4c0d2ab0d3df/geniushubclient/__init__.py#L255-L302

Reply:
@MartinHjelmare, the new client:
 - accept(s) an (optional) aiohttp session - yes, but with the above caveat
 - (optionally) accept an asyncio event loop to use - no, as it doesn't use a loop internally

AFAICT, the PR has passed all tests, and only requires your sign-off on this requested change.  I cannot mark anything as resolved, because I did not submit the original PR.

Would you have a look at it for me?